agent/tool-scripts/bpftrace

```diff
@@ -0,0 +1,123 @@
+#!/bin/bash
+# -*- mode: shell-script; indent-tabs-mode: t; sh-basic-offset: 8; sh-indentation: 8; sh-indent-for-case-alt: + -*-
+# The following script runs a bpftrace program with the program supplied as argument in pbench-register-tool
+# pbench-register-tool --name=bpftrace -- --bpf_prog=sample_prog.bt
+
+script_path=`dirname $0`
+script_name=`basename $0`
+pbench_bin="`cd ${script_path}/..; /bin/pwd`"
+
+# source the base script
+. "$pbench_bin"/base
+
+# Perftool scripts must provide the following functions
+# 1) Install the tool
+# 2) Start data collection
+# 3) Stop data collection
+# 4) post-process the data
+
+# Defaults
+tool=bpftrace
+tool_bin=/usr/local/bin/$tool
+group=default
+dir=""
+mode=""
+iteration="1"
+options="none"
+interval=5
+
+function usage() {
+	echo "
+	USAGE:
+	$script_name: you specified an invalid option
+	The following options are available:
+	--install,			          install this perf tool
+	--start|stop|postprocess	start/stop/post-process the data collection
+	--dir=str			            directory to store data collection (required)
+    --bpf_prog
+	"
+}
+
+
+# Process options and arguments
+opts=$(getopt -q -o idp --longoptions "bpf_prog:,dir:,start,stop,install,postprocess" -n "getopt.sh" -- "$@");
+if [ $? -ne 0 ]; then
+	printf "\n"
+	printf "$script_name: you specified an invalid option\n\n"
+	printf "The following options are available: \n\n"
+	printf -- "\t--install,			install this perf tool\n"
+	printf "\n"
+	printf -- "\t--start|stop|postprocess	start/stop/post-process the data collection\n"
+	printf -- "\t--dir=str			directory to store data collection (required)\n"
+    printf -- "\t--bpf_prog			provide program to run for bpftrace\n"
+fi
+
+eval set -- "$opts";
+while true; do
+	case "$1" in
+		--install)
+		mode="install"
+		shift;
+		;;
+		--start)
+		mode="start"
+		shift;
+		;;
+		--stop)
+		mode="stop"
+		shift;
+		;;
+		--postprocess)
+		mode="postprocess"
+		shift;
+		;;
+        --bpf_prog)
+		shift;
+		if [ -n "$1" ]; then
+			bpf_prog="$1"
+			shift
+		fi
+        ;;
+		-d|--dir)
+		shift;
+		if [ -n "$1" ]; then
+			dir="$1"
+			shift
+		fi
+		;;
+		--)
+		shift;
+		break;
+		;;
+		*)
+		break;
+	esac
+done
+
+tool_cmd="$tool_bin $bpf_prog"
+tool_dir="$dir/tools-$group"
+tool_output_dir=$tool_dir/$tool # all tools keep data in their tool specific dir
+case "$mode" in
+	install)
+	if [ ! -e $tool_bin ]; then
+		error_log "$tool_bin not found, exiting"
+		exit 1
+	fi
+	;;
+	start)
+	mkdir -p $tool_output_dir
+	echo "$script_name: running $tool_cmd"
+	tool_stdout_file=$tool_output_dir/$tool-stdout.txt
+	tool_stderr_file=$tool_output_dir/$tool-stderr.txt
+	tool_cmd_file="$tool_output_dir/$tool.cmd"
+    echo "Running"
+	$tool_cmd  > "$tool_stdout_file" 2> "$tool_stderr_file" &
+    wait
+	;;
+	stop)
+    echo "Killing the process"
+	kill -SIGINT $(pidof bpftrace)
```

Comment:
Instead of using `pidof`, at line 115, can you capture the pid of the tool into a file, like:

```
$tool_cmd  > "$tool_stdout_file" 2> "$tool_stderr_file" &
echo ${!} > ${pbench_tmp}/${tool}.pid
wait
```

And then kill it using:

```
kill -SIGINT $(cat ${pbench_tmp}/${tool}.pid)
```

Or something like the above that works.  Can you find an example of this pattern in the `sar` tool?

Reply:
What's purpose would that serve?